### ISSUE TITLE
Kokkos not Compatible with Big Endian Machines?

### ISSUE THREAD

**Issue by stanmoore1** [kokkos/kokkos#1235]
For example, I don't think code like this in `Kokkos_Rendezvous.cpp` works the same for little and big endian. This appears to be causing the deadlock in #1234 on a big endian machine (BG/Q), because `spinwait_until_equal` never exits when there are more than two OpenMP threads. 

https://github.com/kokkos/kokkos/blob/master/core/src/impl/Kokkos_Rendezvous.cpp#L146

```c++
      // This thread sets its designated byte.
      //   ( rank % size_byte ) +
      //   ( ( rank / size_byte ) * size_byte * size_mem_cycle ) +
      //   ( sync_offset * size_byte )
      const int offset = ( rank & mask_byte )
                       + ( ( rank & ~mask_byte ) << shift_mem_cycle )
                       + ( sync_offset << shift_byte );

      // All of this thread's previous memory stores must be complete before
      // this thread stores the step value at this thread's designated byte
      // in the shared synchronization array.

      Kokkos::memory_fence();

      ((volatile int8_t*) buffer)[ offset ] = int8_t( step );
```

**Comment by stanmoore1** [comment]
For example, here is a simple C++ program that mimics the code in `Kokkos_Rendezvous.cpp` that shows the difference between little and big endian. For this case, Kokkos is expecting a value of 256, and if it doesn't get that it leads to deadlock:

```c++
#include <stdio.h>
#include <stdint.h>

int main()
{
  int64_t d[1024];

  for (int i=0; i<1024; i++)
    d[i] = 0.0;

  int8_t* ptr = (int8_t*) &d[0];
  ptr[73] = int8_t(1);

  printf("value %ld\n",d[9]);

  return 0;
}
```

```
Little Endian: value 256
Big Endian: value 281474976710656
```

I tried on multiple compilers on BG/Q just to make sure it wasn't a compiler bug.

**Comment by stanmoore1** [comment]
I think this is the root cause of Issue #995 and #1234.

**Comment by stanmoore1** [comment]
This issue is at least partially addressed by PR #1237. There may be other places in Kokkos that aren't big endian compatible, but the OpenMP unit tests do pass with PR #1237.

**Comment by dsunder** [comment]
Fixed by #1240
